### PR TITLE
[3/N] Achieve fault tolerance at the DP level

### PR DESCRIPTION
**Pull request by ympcMark**
## Motivation
The previous work of this PR only implemented fault tolerance at the EP level, but not at the DP level; we achieved this functionality by maintaining worker information through communication between the `Scheduler`, `TokenizerManager`, and `DataParallelController`.

## Modifications
`io_struct.py`: Define scheduler status information

`scheduler.py`: Send status information to TokenizerManager.

`tokenizer_manager.py`: Write a handler function to accept this type of information and send information to DataParallelController.

`data_parallel_controller.py`: Write a handler function to accept this type of information and maintain worker information.

## Accuracy Tests
added a new unit test `test/srt/ep/test_mooncake_ep_small.py`

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

**Comment by UNIDY2002** [comment]
TODOs before review:
1. Support multiple nodes
2. Retry mechanism for failing DP ranks
3. Make Mooncake EP's a2a implementation runnable on the CI
4. Comprehensive tests

File: python/sglang/srt/distributed/parallel_state.py [hunk 1]
```diff
@@ -1311,10 +1334,23 @@ def init_model_parallel_group(
...
         pynccl_use_current_stream=pynccl_use_current_stream,
         torch_compile=torch_compile,
+        active_ranks=active_ranks,
+        active_ranks_cpu=active_ranks_cpu,
     )
 
 
 _TP: Optional[GroupCoordinator] = None
+_TP_ACTIVE_RANKS: Optional[torch.Tensor] = None
+_TP_ACTIVE_RANKS_CPU: Optional[torch.Tensor] = None
```
Comment (HanHan009527) [inline python/sglang/srt/distributed/parallel_state.py]: I'm thinking, would it be more reasonable to rename the elastic ep module to elstic and group together concepts like _TP_ACTIVE_RANKS and _TP_ACTIVE_RANKS_CPU as much as possible?

Comment (HanHan009527) [reply python/sglang/srt/distributed/parallel_state.py]: and  get set functions operate on them

Comment (HanHan009527) [reply python/sglang/srt/distributed/parallel_state.py]: Make it less intrusive to the original logic

Comment (UNIDY2002) [reply python/sglang/srt/distributed/parallel_state.py]: Sounds reasonable

File: python/sglang/srt/managers/data_parallel_controller.py [hunk 2]
```diff
@@ -158,6 +159,7 @@ def __init__(
         # Launch data parallel workers
         self.scheduler_procs = []
         self.workers: List[zmq.Socket] = [None] * server_args.dp_size
+        self.status: List[int] = [1] * server_args.dp_size
```
Comment (ShangmingCai) [inline python/sglang/srt/managers/data_parallel_controller.py]: What is the range of possible values ​​for this status?

File: python/sglang/srt/managers/scheduler_dp_attn_mixin.py [hunk 3]
```diff
@@ -66,6 +68,15 @@ def all_gather(self, device, group: torch.distributed.ProcessGroup):
             local_info_tensor,
             group=group,
         )
+        if device == "cpu":
+            tp_active_ranks = get_tp_group().active_ranks_cpu
+        else:
+            tp_active_ranks = get_tp_group().active_ranks
+        global_info_tensor.view(-1, 6)[tp_active_ranks == 0, :] = torch.tensor(
+            [0, 1, 0, 0, 1, ForwardMode.IDLE.value],
```
Comment (ShangmingCai) [inline python/sglang/srt/managers/scheduler_dp_attn_mixin.py]: The code here seems to have poor readability.

File: python/sglang/srt/managers/data_parallel_controller.py (hunk 2 above)
Comment (ShangmingCai) [reply python/sglang/srt/managers/data_parallel_controller.py]: Should be better to use an `enum class` here, I think`x== 1` is not easy to understand.

**Review by ShangmingCai — COMMENTED** [review]
Make sense. We need to disable failed ep rank and dp rank to achieve fault tolerance during large-scale deployment. However, Scheduler/Tokenizer Manager/DP changes should be reviewed carefully by @hnyls2002. Please ping him in the Slack.

File: python/sglang/srt/managers/data_parallel_controller.py (hunk 2 above)
Comment (ympcMark) [reply python/sglang/srt/managers/data_parallel_controller.py]: status contains only 0 and 1, representing whether the worker is available

Comment (ympcMark) [reply python/sglang/srt/managers/data_parallel_controller.py]: shall we use bool ?

Comment (ShangmingCai) [reply python/sglang/srt/managers/data_parallel_controller.py]: yeah, bool sounds better

File: test/srt/ep/test_mooncake_ep_small.py [hunk 4]
```diff
@@ -70,16 +71,36 @@ def test_gsm8k(self):
...
+        "--moe-dense-tp-size",
+        "1",
+        "--enable-dp-lm-head",
+        "--disable-custom-all-reduce",
+        "--enable-eplb",
+        "--ep-num-redundant-experts",
+        "72",
     ]
 
+    def test_gsm8k_fault_1(self):
```
Comment (ch-wan) [inline test/srt/ep/test_mooncake_ep_small.py]: move these tests to the base class

**Review by ch-wan — APPROVED** [review]
LGTM in general. I only have one minor comment.

File: python/sglang/srt/managers/scheduler_dp_attn_mixin.py (hunk 3 above)
Comment (UNIDY2002) [reply python/sglang/srt/managers/scheduler_dp_attn_mixin.py]: Fixed

File: test/srt/ep/test_mooncake_ep_small.py (hunk 4 above)
Comment (UNIDY2002) [reply test/srt/ep/test_mooncake_ep_small.py]: Fixed
